### ISSUE TITLE
ABI version not set for json gem

### ISSUE THREAD

**Issue by nirvdrum** [oracle/truffleruby#2386]
The `TRUFFLERUBY_ABI_VERSION` value is not being set correctly when installing _json_ from a gem:

```
nirvdrum@melchior-linux ~/d/w/ruby-benchmarks> gem install json
Fetching json-2.5.1.gem
Building native extensions. This could take a while...
Successfully installed json-2.5.1
1 gem installed

nirvdrum@melchior-linux ~/d/w/ruby-benchmarks> ruby -r json -e 'p :hi'
warning: The native extension at /home/nirvdrum/dev/workspaces/truffleruby-ws/graal/sdk/mxbuild/linux-amd64/GRAALVM_92F985B265_JAVA11/graalvm-92f985b265-java11-21.2.0-dev/languages/ruby/lib/gems/gems/json-2.5.1/lib/json/ext/parser.so has a different ABI version: nil than the running TruffleRuby: "2.7.3.5"
```

It is being generated in the various _Makefiles_:

```
nirvdrum@melchior-linux ~/d/w/t/g/s/m/l/G/g/l/r/l/g/g/json-2.5.1 ((4f4d2762…))> rg TRUFFLERUBY_ABI_VERSION
ext/json/ext/parser/Makefile
64:DEFS     = -DTRUFFLERUBY_ABI_VERSION=2.7.3.5

ext/json/ext/generator/Makefile
64:DEFS     = -DTRUFFLERUBY_ABI_VERSION=2.7.3.5

ext/json/Makefile
64:DEFS     = -DTRUFFLERUBY_ABI_VERSION=2.7.3.5
```

**Comment by nirvdrum** [comment]
I'm seeing a similar issue with [grpc](https://github.com/Shopify/grpc) (branch: _truffleruby-1.30.1_). Since the feature isn't broadly broken, would you prefer a separate issue is filed for each affected extension?

**Comment by eregon** [comment]
I can reproduce for `json`, `lib/json/ext/generator.so` has the version but `lib/json/ext/parser.so` seems to have almost no symbols, investigating.

**Comment by eregon** [comment]
Ugh, it seems the preprocessor fails but then nothing really notices the failure and it keeps installing:
```
$ jt gem i -V json

compiling parser.c
lib/truffle/truffle/cext_preprocessor.rb:99:in `block in patch': undefined method `lstrip' for nil:NilClass (NoMethodError)
	from lib/truffle/truffle/cext_preprocessor.rb:98:in `each'
	from lib/truffle/truffle/cext_preprocessor.rb:98:in `patch'
	from lib/cext/preprocess.rb:15:in `<main>'
```

**Comment by eregon** [comment]
About `grpc`, the problem is likely that it strips the symbols.
https://github.com/Shopify/grpc/blob/9bfe26e03b8122163e10acaa563e52616244a713/src/ruby/ext/grpc/extconf.rb#L129-L137

So:
```
$ nm /home/eregon/.rubies/truffleruby-dev/lib/gems/bundler/gems/grpc-9bfe26e03b81/src/ruby/lib/grpc/grpc_c.so | grep 'rb_tr'
nothing
$ nm -D /home/eregon/.rubies/truffleruby-dev/lib/gems/bundler/gems/grpc-9bfe26e03b81/src/ruby/lib/grpc/grpc_c.so | grep rb_tr
                 U rb_tr_cext
                 U rb_tr_id2sym
                 U rb_tr_longwrap
                 U rb_tr_sym2id
                 U rb_tr_unwrap
                 U rb_tr_wrap
```

**Comment by eregon** [comment]
Just removing the `strip` does not seem enough unfortunately, the symbols are still undefined:
```
$ nm /home/eregon/.rubies/truffleruby-dev/lib/gems/gems/grpc-1.36.0.dev/src/ruby/lib/grpc/grpc_c.so | grep rb_tr
                 U rb_tr_cext
                 U rb_tr_id2sym
                 U rb_tr_longwrap
                 U rb_tr_sym2id
                 U rb_tr_unwrap
                 U rb_tr_wrap
```

**Comment by eregon** [comment]
Actually the `strip` is fine, the issue is due to `-Wl,--version-script=` https://github.com/Shopify/grpc/blob/9bfe26e03b8122163e10acaa563e52616244a713/src/ruby/ext/grpc/extconf.rb#L96 which hides most symbols.

**Comment by eregon** [comment]
https://github.com/eregon/grpc/commit/d400bb98a9c25fb0d1f534a7eb9e7946846d0305 fixes it by explictly keeping the `rb_tr_abi_version` symbol. Then `require 'grpc'` works fine.

There is also the possibility to skip https://github.com/Shopify/grpc/blob/9bfe26e03b8122163e10acaa563e52616244a713/src/ruby/ext/grpc/extconf.rb#L96 entirely, but that seems a bigger change.

**Comment by nirvdrum** [comment]
I'm a bit worried this is going to continue happening with arbitrary gems and each will require a one-off fix. "Fixing" it in each gem can be problematic if a project can't or won't upgrade to a version of the gem that pulls in the fix. E.g., some gems have infrequent release cycles and having to point at a git branch may not be allowed.

Can we go about this another way? It's a terribly naive solution, but could we write a text file that sits next to the extension that has the ABI version? Then part of our library load process would be reading that ABI file if it exists. Something along those lines would be more resilient to subtle differences in project _Makefiles_.

**Comment by eregon** [comment]
`grpc` is the only gem so far that needs upstream modification for this, and it (currently) needs modifications to compile on TruffleRuby anyway.
So I think it's good enough for now. I'm happy to reconsider if there are other gems that would need upstream fixes.

An extra file seems difficult to produce next to the `.so` and much less reliable as it doesn't indicate which version of the headers was used. If the current approach doesn't work well enough, we could try to add an extra `.c` file which includes a function with the ABI version, compile it, and add it to the `.so`, but in the case of grpc it wouldn't solve the issue.

**Comment by eregon** [comment]
The issue is fixed for `json` in a8b7abcf89458611e53107b4cb493abc193bc062, and for `grpc` it needs https://github.com/eregon/grpc/commit/d400bb98a9c25fb0d1f534a7eb9e7946846d0305.
Thanks for the report!
